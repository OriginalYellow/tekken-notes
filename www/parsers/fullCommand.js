/* eslint-disable no-unused-expressions */
import * as R from 'ramda'
import * as RA from 'ramda-adjunct'
import * as A from 'arcsecond'
// import consecutive from './scratch-019'
import consecutive from '../scratch/scratch-019'

const whitespaceSurrounded = parser =>
  A.between(A.optionalWhitespace)(A.optionalWhitespace)(parser)

const betweenParens = parser =>
  A.between(A.char('('))(A.char(')'))(parser)

const toTypedObj = type => val => ({ type, val })
const unnest2 = R.pipe(R.unnest, R.unnest)
const lastIsNull = R.pipe(R.last, RA.isNull)
const sepByOr = A.sepBy1(whitespaceSurrounded(A.str('or')))

const conditionString = terminator => stringOrArray => A.regex(new RegExp(
  `^${
  RA.isString(stringOrArray) ? stringOrArray : stringOrArray[0]
  }(?=${terminator})`,
  'i'
)).map(R.ifElse(
  R.always(RA.isString(stringOrArray)),
  R.identity,
  R.always(stringOrArray[1])
)).map(R.toLower)

const prefixConditionStrings = R.map(conditionString('\\s'))

const inputAtom = whitespaceSurrounded(
  A.sequenceOf([
    A.regex(/^[1234]/),
    A.lookAhead(A.choice([
      A.regex(/^\s*(?:[,+~(*]|or|$)/),
      A.endOfInput
    ]))
  ])
)
  .map(R.head)
  .map(toTypedObj('atom'))

const inputAtomOperator = A.char('+')
  .map(toTypedObj('operator'))

const inputMoleculeBody = A.sequenceOf([
  inputAtom,
  A.possibly(consecutive(A.sequenceOf([
    inputAtomOperator,
    inputAtom
  ]))),
  A.lookAhead(A.choice([
    A.regex(/^\s*(?:[(,~*]|or|$)/),
    A.endOfInput
  ]))
])
  .map(unnest2)
  .map(R.init)
  .map(R.ifElse(
    lastIsNull,
    R.init,
    R.identity
  ))

const inputMoleculePrefixCondition = A.possibly(A.choice(prefixConditionStrings([
  'when hit'
])))

const inputMoleculeSuffixCondition = A.possibly(whitespaceSurrounded(A.choice([
  A.char('*')
])))

const inputMolecule = A.namedSequenceOf([
  [
    'prefixCondition',
    inputMoleculePrefixCondition
  ],
  [
    'body',
    inputMoleculeBody
  ],
  [
    'suffixCondition',
    inputMoleculeSuffixCondition
  ]
]).map(toTypedObj('molecule'))

const inputMoleculeOperator = A.choice([
  whitespaceSurrounded(A.char(',')),
  whitespaceSurrounded(A.char('~'))
]).map(toTypedObj('operator'))

const inputMolecules = A.sequenceOf([
  inputMolecule,
  A.possibly(consecutive(A.sequenceOf([
    inputMoleculeOperator,
    inputMolecule
  ]))),
  A.lookAhead(A.choice([
    A.regex(/^\s*(?:or|[(]|$)/),
    A.endOfInput
  ]))
])
  .map(unnest2)
  .map(R.init)
  .map(R.ifElse(
    lastIsNull,
    R.init,
    R.identity
  ))

const stancePrefixCondition = A.choice(prefixConditionStrings([
  'dmn',
  'rss'
])).map(toTypedObj('stance'))

const otherPrefixCondition = A.choice(prefixConditionStrings([
  ['(in rage|rage|during rage)', 'rage'],
  'jump'
])).map(toTypedObj('other'))

const prefixCondition = whitespaceSurrounded(A.choice([
  stancePrefixCondition,
  otherPrefixCondition
]))

const prefixConditions = A.possibly(A.many(prefixCondition))

const suffixConditionStrings = R.map(conditionString('[)]'))

const getSuffixCondition = R.compose(
  A.possibly,
  whitespaceSurrounded,
  betweenParens,
  A.choice,
  suffixConditionStrings
)

const suffixCondition = getSuffixCondition([
  'cancel',
  'close',
  'far'
])

// const alternativeCommand = A.namedSequenceOf([
//   ['prefixConditions', prefixConditions.errorMap((x) => {
//     x // ?
//     return x
//   }).map((x) => {
//     x // ?
//     return x
//   })],
//   ['inputMolecules', inputMolecules.errorMap((x) => {
//     x // ?
//     return x
//   }).map((x) => {
//     x // ?
//     return x
//   })],
//   ['suffixCondition', suffixCondition.errorMap((x) => {
//     x // ?
//     return x
//   }).map((x) => {
//     x // ?
//     return x
//   })]
// ])

const alternativeCommand = A.namedSequenceOf([
  ['prefixConditions', prefixConditions],
  ['inputMolecules', inputMolecules],
  ['suffixCondition', suffixCondition]
])

const gotStringRegex = /got '(.*)(?='$)/
const indexStringRegex = /position (\d*)/

export const fullCommand = sepByOr(
  alternativeCommand
)
  .errorChain(({ error }) => {
    if (~error.substring('Expecting to match at least one separated value')) {
      return alternativeCommand
    }

    return A.fail(error)
  })
  .errorMap((error) => {
    const index = error.match(indexStringRegex) ? error.match(indexStringRegex)[1] : null

    if (~error.indexOf('but got end of input')) {
      return `expected more text at position ${index}`
    }

    const gotString = error.match(gotStringRegex) ? error.match(gotStringRegex)[1] : null
    return `"...${gotString}" at position ${index} is invalid`
  })

export default fullCommand
