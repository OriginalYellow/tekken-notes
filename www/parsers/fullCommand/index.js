export { parse } from './parse'
export { compileToProps } from './compileToProps'
