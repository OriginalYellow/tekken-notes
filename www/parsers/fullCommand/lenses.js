import * as L from 'partial.lenses'
import {
  SUFFIX_CONDITION,
  PREFIX_CONDITION,
  PREFIX_CONDITIONS,
  INPUT_MOLECULES
} from './constants'

export const Model = {
  prefixConditions: [PREFIX_CONDITIONS, L.elems],
  suffixCondition: [SUFFIX_CONDITION],
  inputMolecules: [INPUT_MOLECULES, L.elems]
}

export const InputMolecule = {
  body: ['val', 'body', L.elems],
  prefixCondition: ['val', PREFIX_CONDITION],
  suffixCondition: ['val', SUFFIX_CONDITION]
}
