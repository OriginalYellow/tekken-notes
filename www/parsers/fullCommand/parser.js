/* eslint-disable func-call-spacing */
/* eslint-disable no-unused-expressions */
import * as R from 'ramda'
import * as RA from 'ramda-adjunct'
import * as A from 'arcsecond'
import consecutive from './consecutive'
import * as U from './util'
import {
  ATOM,
  MOLECULE,
  OPERATOR,
  PREFIX_CONDITION,
  STANCE_CONDITION,
  OTHER_CONDITION
} from './types'

R.if = R.ifElse(R.__, R.__, R.identity)

export const inputAtomRegex = /^[1234]/
const inputAtomTerminatorRegex = /^\s*(?:[,+~(*]|or|$)/
const inputMoleculeBodyTerminatorRegex = /^\s*(?:[(,~*]|or|$)/
const inputMoleculesTerminatorRegex = /^\s*(?:or|[(]|$)/
const alternativeCommandTerminatorRegex = /^\s*(?=or)/
const gotStringRegex = /got '(.*)(?='$)/
const indexStringRegex = /position (\d*)/

const whitespaceSurrounded = parser =>
  A.between(A.optionalWhitespace)(A.optionalWhitespace)(parser)

const betweenParens = parser =>
  A.between(A.char('('))(A.char(')'))(parser)

const sepByOr = A.sepBy1(whitespaceSurrounded(A.str('or')))

const conditionString = terminator => stringOrArray => A.regex(new RegExp(
  `^${
  RA.isString(stringOrArray) ? stringOrArray : stringOrArray[0]
  }(?=${terminator})`,
  'i'
)).map(R.ifElse(
  R.always(RA.isString(stringOrArray)),
  R.identity,
  R.always(stringOrArray[1])
)).map(R.toLower)

const prefixConditionStrings = R.map(conditionString('\\s'))

const inputAtom = whitespaceSurrounded(
  A.sequenceOf([
    A.regex(inputAtomRegex),
    A.lookAhead(A.choice([
      A.regex(inputAtomTerminatorRegex),
      A.endOfInput
    ]))
  ])
)
  .map(R.head)
  .map(U.toTypedObj(ATOM))

const inputAtomOperator = A.char('+')
  .map(U.toTypedObj(OPERATOR))

const inputMoleculeBody = A.sequenceOf([
  inputAtom,
  A.possibly(consecutive(A.sequenceOf([
    inputAtomOperator,
    inputAtom
  ]))),
  A.lookAhead(A.choice([
    A.regex(inputMoleculeBodyTerminatorRegex),
    A.endOfInput
  ]))
])
  .map(RA.flattenDepth(2))
  .map(R.init)
  .map(R.if(
    U.lastIsNull,
    R.init
  ))

const inputMoleculePrefixCondition = A.possibly(A.choice(prefixConditionStrings([
  'when hit'
])).map(U.toTypedObj(PREFIX_CONDITION)))

const inputMoleculeSuffixCondition = A.possibly(whitespaceSurrounded(A.choice([
  A.char('*')
])))

const inputMolecule = A.namedSequenceOf([
  [
    PREFIX_CONDITION,
    inputMoleculePrefixCondition
  ],
  [
    'body',
    inputMoleculeBody
  ],
  [
    'suffixCondition',
    inputMoleculeSuffixCondition
  ]
]).map(U.toTypedObj(MOLECULE))

const inputMoleculeOperator = A.choice([
  whitespaceSurrounded(A.char(',')),
  whitespaceSurrounded(A.char('~'))
]).map(U.toTypedObj(OPERATOR))

const inputMolecules = A.sequenceOf([
  inputMolecule,
  A.possibly(consecutive(A.sequenceOf([
    inputMoleculeOperator,
    inputMolecule
  ]))),
  A.lookAhead(A.choice([
    A.regex(inputMoleculesTerminatorRegex),
    A.endOfInput
  ]))
])
  .map(RA.flattenDepth(2))
  .map(R.init)
  .map(R.if(
    U.lastIsNull,
    R.init
  ))

const stancePrefixCondition = A.choice(prefixConditionStrings([
  'dmn',
  'rss'
])).map(U.toTypedObj(STANCE_CONDITION))

const otherPrefixCondition = A.choice(prefixConditionStrings([
  ['(in rage|rage|during rage)', 'rage'],
  'jump'
])).map(U.toTypedObj(OTHER_CONDITION))

const prefixCondition = whitespaceSurrounded(A.choice([
  stancePrefixCondition,
  otherPrefixCondition
]))

const prefixConditions = A.possibly(A.many(prefixCondition))

const suffixConditionStrings = R.map(conditionString('[)]'))

const getSuffixCondition = R.compose(
  whitespaceSurrounded,
  betweenParens,
  A.choice,
  suffixConditionStrings
)

const suffixCondition = getSuffixCondition([
  'cancel',
  'close',
  'far'
])

const alternativeCommand = A.namedSequenceOf([
  ['prefixConditions', prefixConditions],
  ['inputMolecules', inputMolecules],
  [
    'suffixCondition',
    A.choice([
      suffixCondition,
      A.endOfInput,
      A.regex(alternativeCommandTerminatorRegex)
    ]).map(R.if(
      RA.isNilOrEmpty,
      R.always(null)
    ))
  ]
])

export const fullCommand = sepByOr(
  alternativeCommand
)
  .errorChain(({ error }) => {
    if (~error.indexOf('Expecting to match at least one separated value')) {
      return alternativeCommand
    }

    return A.fail(error)
  })
  .errorMap((error) => {
    const index = error.match(indexStringRegex) ? error.match(indexStringRegex)[1] : null

    if (~error.indexOf('but got end of input')) {
      return `expected more text at position ${index}`
    }

    const gotString = error.match(gotStringRegex) ? error.match(gotStringRegex)[1] : null
    return `"...${gotString}" at position ${index} is invalid`
  })
