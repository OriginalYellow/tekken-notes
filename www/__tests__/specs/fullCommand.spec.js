// import * as RA from 'ramda-adjunct'
// import { fullCommand } from '../../scratch/scratch-018.js'
import { fullCommand } from '../../parsers/fullCommand/parser'

it('"1 or 4, 2" parses correctly', () => {
  const tree = fullCommand.run('1 or 4, 2')
  expect(tree).toMatchSnapshot()
})

it('"1+3" successfully parses', () => {
  expect(fullCommand.run('1+3').error).toBeFalsy()
})

it('"1+3, when hit 2+1" successfully parses', () => {
  expect(fullCommand.run('1+3, when hit 2+1').error).toBeFalsy()
})

it('"1+3*, when hit 2+1*" successfully parses', () => {
  fullCommand.run('1+3*, when hit 2+1*') // ?
  expect(fullCommand.run('1+3*, when hit 2+1 * ').error).toBeFalsy()
})

it('"3*" successfully parses', () => {
  expect(fullCommand.run('3*').error).toBeFalsy()
})

it('"1+3, when hit 2+1" parses correctly', () => {
  const tree = fullCommand.run('1+3, when hit 2+1')
  expect(tree).toMatchSnapshot()
})

it('"1+3 (cancel) or 1" parses correctly', () => {
  const tree = fullCommand.run('1+3 (cancel) or 1')
  expect(tree).toMatchSnapshot()
})

it('"1+3 (cancel)" parses correctly', () => {
  const tree = fullCommand.run('1+3 (cancel)')
  expect(tree).toMatchSnapshot()
})

it('"dmn in rage 2+4, 1~4+3, 1 or 1" parses correctly', () => {
  const tree = fullCommand.run('dmn in rage 2+4, 1~4+3, 1 or 1')
  expect(tree).toMatchSnapshot()
})

it('"dmn in rage 2, 1+4 or 1" parses correctly', () => {
  const tree = fullCommand.run('dmn in rage 2, 1+4 or 1')
  expect(tree).toMatchSnapshot()
})

it('"rss in rage 2, 1 or 1" parses correctly', () => {
  const tree = fullCommand.run('rss in rage 2, 1 or 1')
  expect(tree).toMatchSnapshot()
})

it('"in rage DMN 2, 1" parses correctly', () => {
  const tree = fullCommand.run('in rage DMN 2, 1')
  expect(tree).toMatchSnapshot()
})

it('"jump 2, 1" parses correctly', () => {
  const tree = fullCommand.run('jump 2, 1')
  expect(tree).toMatchSnapshot()
})

it('" 2, 1" parses correctly', () => {
  const tree = fullCommand.run(' 2, 1')
  expect(tree).toMatchSnapshot()
})

it('"DMN jump rage 2, 1" parses correctly', () => {
  const tree = fullCommand.run('DMN jump rage 2, 1')
  expect(tree).toMatchSnapshot()
})

it('"1+5" should error', () => {
  expect(fullCommand.run('1+5').error).toBeTruthy()
  expect(fullCommand.run('1+5').error).toMatchSnapshot()
})

it('"4+" should error', () => {
  expect(fullCommand.run('4+').error).toBeTruthy()
  expect(fullCommand.run('4+').error).toMatchSnapshot()
})

it('"4~, 2" should error', () => {
  expect(fullCommand.run('4~, 2').error).toBeTruthy()
  expect(fullCommand.run('4~, 2').error).toMatchSnapshot()
})

it('"1  + (cancel) or 1" should error', () => {
  expect(fullCommand.run('1  + (cancel) or 1').error).toBeTruthy()
  expect(fullCommand.run('1  + (cancel) or 1').error).toMatchSnapshot()
})

it('"rssin rage 2, 1 or 1" should error', () => {
  expect(fullCommand.run('rssin rage 2, 1 or 1').error).toBeTruthy()
  expect(fullCommand.run('rssin rage 2, 1 or 1').error).toMatchSnapshot()
})

it('"1, 3 in rage" should error', () => {
  expect(fullCommand.run('1, 3 in rage').error).toBeTruthy()
  expect(fullCommand.run('1, 3 in rage').error).toMatchSnapshot()
})

it('"1+5, asdf" should error', () => {
  expect(fullCommand.run('1+5, asdf').error).toBeTruthy()
  expect(fullCommand.run('1+5, asdf').error).toMatchSnapshot()
})

it('"1+4, asdf" should error', () => {
  expect(fullCommand.run('1+4, asdf').error).toBeTruthy()
  expect(fullCommand.run('1+4, asdf').error).toMatchSnapshot()
})

it('"adf 1, 3 in rage" should error', () => {
  expect(fullCommand.run('adf 1, 3 in rage').error).toBeTruthy()
  expect(fullCommand.run('adf 1, 3 in rage').error).toMatchSnapshot()
})

it('"dmn in rage 2, 1+4 asdf or 1" should error', () => {
  expect(fullCommand.run('dmn in rage 2, 1+4 asdf or 1').error).toBeTruthy()
  expect(fullCommand.run('dmn in rage 2, 1+4 asdf or 1').error).toMatchSnapshot()
})

it('"dmn in rage 2,, 1+4 or 1" should error', () => {
  expect(fullCommand.run('dmn in rage 2,, 1+4 or 1').error).toBeTruthy()
  expect(fullCommand.run('dmn in rage 2,, 1+4 or 1').error).toMatchSnapshot()
})

it('"in rage " should error', () => {
  expect(fullCommand.run('in rage ').error).toBeTruthy()
  expect(fullCommand.run('in rage ').error).toMatchSnapshot()
})

it('"1 (" should error', () => {
  expect(fullCommand.run('1 (').error).toBeTruthy()
  expect(fullCommand.run('1 (').error).toMatchSnapshot()
})

it('"1 (asdf)" should error', () => {
  expect(fullCommand.run('1 (asdf)').isError).toBeTruthy()
  expect(fullCommand.run('1 (asdf)').error).toMatchSnapshot()
})
