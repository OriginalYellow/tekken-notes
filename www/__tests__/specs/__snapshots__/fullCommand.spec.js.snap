// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`" 2, 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 5,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"1  + (cancel) or 1" should error 1`] = `"\\"...+ (ca...\\" at position 3 is invalid"`;

exports[`"1 (" should error 1`] = `"expected more text at position 3"`;

exports[`"1 (asdf)" should error 1`] = `"\\"...asdf)...\\" at position 3 is invalid"`;

exports[`"1 or 4, 2" parses correctly 1`] = `
Object {
  "data": null,
  "index": 9,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "4",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"1+3 (cancel) or 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 17,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "3",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": "cancel",
      },
      "type": "alt command",
    },
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"1+3 (cancel)" parses correctly 1`] = `
Object {
  "data": null,
  "index": 12,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "3",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": "cancel",
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"1+3, when hit 2+1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 17,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "3",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": Object {
                "type": "prefix condition",
                "val": "when hit",
              },
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"1+4, asdf" should error 1`] = `"\\"..., asd...\\" at position 3 is invalid"`;

exports[`"1+5" should error 1`] = `"\\"...+5...\\" at position 1 is invalid"`;

exports[`"1+5, asdf" should error 1`] = `"\\"...+5, a...\\" at position 1 is invalid"`;

exports[`"1, 3 in rage" should error 1`] = `"\\"..., 3 i...\\" at position 1 is invalid"`;

exports[`"4+" should error 1`] = `"\\"...+...\\" at position 1 is invalid"`;

exports[`"4~, 2" should error 1`] = `"\\"...~, 2...\\" at position 1 is invalid"`;

exports[`"DMN jump rage 2, 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 18,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "stance condition",
            "val": "dmn",
          },
          Object {
            "type": "other condition",
            "val": "jump",
          },
          Object {
            "type": "other condition",
            "val": "rage",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"DMN jump rage U/B+2, 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 22,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "U/B",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "stance condition",
            "val": "dmn",
          },
          Object {
            "type": "other condition",
            "val": "jump",
          },
          Object {
            "type": "other condition",
            "val": "rage",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"DMN jump rage u/b+2, 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 22,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "u/b",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "stance condition",
            "val": "dmn",
          },
          Object {
            "type": "other condition",
            "val": "jump",
          },
          Object {
            "type": "other condition",
            "val": "rage",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"adf 1, 3 in rage" should error 1`] = `"\\"...adf 1...\\" at position 0 is invalid"`;

exports[`"dmn in rage 2+4, 1~4+3, 1 or 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 30,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "4",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": "~",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "4",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "3",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "stance condition",
            "val": "dmn",
          },
          Object {
            "type": "other condition",
            "val": "rage",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"dmn in rage 2, 1+4 asdf or 1" should error 1`] = `"\\"..., 1+4...\\" at position 13 is invalid"`;

exports[`"dmn in rage 2, 1+4 or 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 23,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
                Object {
                  "type": "operator",
                  "val": "+",
                },
                Object {
                  "type": "atom",
                  "val": "4",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "stance condition",
            "val": "dmn",
          },
          Object {
            "type": "other condition",
            "val": "rage",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"dmn in rage 2,, 1+4 or 1" should error 1`] = `"\\"...,, 1+...\\" at position 13 is invalid"`;

exports[`"in rage " should error 1`] = `"expected more text at position 8"`;

exports[`"in rage DMN 2, 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 16,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "other condition",
            "val": "rage",
          },
          Object {
            "type": "stance condition",
            "val": "dmn",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"jump 2, 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 9,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "other condition",
            "val": "jump",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"rss in rage 2, 1 or 1" parses correctly 1`] = `
Object {
  "data": null,
  "index": 21,
  "isError": false,
  "result": Array [
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "2",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
          Object {
            "type": "operator",
            "val": ",",
          },
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [
          Object {
            "type": "stance condition",
            "val": "rss",
          },
          Object {
            "type": "other condition",
            "val": "rage",
          },
        ],
        "suffix condition": null,
      },
      "type": "alt command",
    },
    Object {
      "body": Object {
        "input molecules": Array [
          Object {
            "type": "molecule",
            "val": Object {
              "body": Array [
                Object {
                  "type": "atom",
                  "val": "1",
                },
              ],
              "prefix condition": null,
              "suffix condition": null,
            },
          },
        ],
        "prefix conditions": Array [],
        "suffix condition": null,
      },
      "type": "alt command",
    },
  ],
}
`;

exports[`"rssin rage 2, 1 or 1" should error 1`] = `"\\"...rssin...\\" at position 0 is invalid"`;
